include/field.hxx

```diff
@@ -95,6 +114,11 @@ class Field {
     if (fieldmesh){
       return fieldmesh;
     } else {
+      // Don't set fieldmesh=mesh here, so that fieldmesh==nullptr until
+      // allocate() is called in one of the derived classes. fieldmesh==nullptr
+      // indicates that some initialization that would be done in the
+      // constructor if fieldmesh was a valid Mesh object still needs to be
+      // done.
```

Comment:
I agree this hasn't ended up being a good implementation, but I don't think we can just return `nullptr` here:
* If `fieldmesh==nullptr` then when the field is allocated, we will set `fieldmesh=bout::globals::mesh`, and since `getMesh()` might be called before the field is allocated, we want to return `bout::globals::mesh`.
* The reason we can't set it here is that we use `fieldmesh==nullptr` to indicate that `allocate()` needs to do some additional set-up that couldn't be done in the constructor because we didn't have a valid mesh then: namely setting, if needed for the child class, `nx`/`ny`/`nz` and default values of `xDirectionType`/`yDirectionType`/`zDirectionType`.

Would be nice for this to be able to require fields to have a valid mesh in the constructor, but that would break the interface too much.

Would it be better to add something like `bool fullyInitialised{false}` to `Field` and check/update that in `allocate()`? Then the intention would be clearer from the variable name and we could set `fieldmesh = bout::globals::mesh` in `getMesh()` too.

Reply:
Ok, looks like we need to do some more thinking about what we should/want to be doing here. Definitely future work though